### PR TITLE
Feature/aql subquery execution block impl execute implementation exec block impl tests

### PR DESCRIPTION
**Pull request by mchacki**
Internal merge.
This will add some tests to ExecutionBlockImpl logic.
It will also implement fullCount and fix an issue on Skip.

NOTE:
There are still 3 disabled tests here, that cannot be green right now.
We need to modify the OutputItemRow first.

Jenkins:
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/7932/

### CODE REVIEW

File: tests/Aql/TestLambdaExecutor.h [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+                          std::unordered_set<RegisterId> registersToClear,
+                          std::unordered_set<RegisterId> registersToKeep,
+                          ProduceCall lambda, SkipCall skipLambda);
+
+  LambdaSkipExecutorInfos() = delete;
+  LambdaSkipExecutorInfos(LambdaSkipExecutorInfos&&) = default;
+  LambdaSkipExecutorInfos(LambdaSkipExecutorInfos const&) = delete;
+  ~LambdaSkipExecutorInfos() = default;
+
+  auto getLambda() const -> ProduceCall const&;
```
Comment (markuspf) [inline tests/Aql/TestLambdaExecutor.h]:
```suggestion
  auto getProduceLambda() const -> ProduceCall const&;
```

File: tests/Aql/ExecutionBlockImplTest.cpp [hunk 2]
```diff
@@ -391,6 +401,925 @@ TEST_F(ExecutionBlockImplTest,
...
+}
+
+TEST_F(ExecutionBlockImplExecuteTest, test_toplevel_softlimit_call) {
+  AqlCall fullCall{};
+  fullCall.softLimit = 20;
+  AqlCallStack stack{std::move(fullCall)};
+  ExecutionNode const* node = nullptr;
+  size_t nrCalls = 0;
+
+  ProduceCall execImpl = [&nrCalls](AqlItemBlockInputRange& input, OutputAqlItemRow& output)
```
Comment (markuspf) [inline tests/Aql/ExecutionBlockImplTest.cpp]: Is this produceCall different from the one above?

File: tests/Aql/ExecutionBlockImplTest.cpp [hunk 3]
```diff
@@ -391,6 +401,925 @@ TEST_F(ExecutionBlockImplTest,
...
+}
+
+TEST_F(ExecutionBlockImplExecuteTest, test_toplevel_hardlimit_call) {
+  AqlCall fullCall{};
+  fullCall.hardLimit = 20;
+  AqlCallStack stack{std::move(fullCall)};
+  ExecutionNode const* node = nullptr;
+  size_t nrCalls = 0;
+
+  ProduceCall execImpl = [&nrCalls](AqlItemBlockInputRange& input, OutputAqlItemRow& output)
```
Comment (markuspf) [inline tests/Aql/ExecutionBlockImplTest.cpp]: Same as above

File: tests/Aql/ExecutionBlockImplTest.cpp [hunk 4]
```diff
@@ -391,6 +401,925 @@ TEST_F(ExecutionBlockImplTest,
...
+    if (nrCalls > 10) {
+      EXPECT_TRUE(false);
+      // This is emergency bailout, we ask way to often here
+      THROW_ARANGO_EXCEPTION(TRI_ERROR_INTERNAL);
+    }
+    // We are not allowed to get called at all.
+    EXPECT_FALSE(true);
+    nrCalls++;
+    NoStats stats{};
+    AqlCall call{};
```
Comment (markuspf) [inline tests/Aql/ExecutionBlockImplTest.cpp]: not reached?

File: tests/Aql/ExecutionBlockImplTest.cpp [hunk 5]
```diff
@@ -391,6 +401,925 @@ TEST_F(ExecutionBlockImplTest,
...
+      LOG_DEVEL << got;
+      if (call > maxCall + 2) {
+        // Security bailout to avoid infinite loops
+        THROW_ARANGO_EXCEPTION(TRI_ERROR_INTERNAL);
+      }
+    }
+  }
+};
+
+// Asserts if AqlCalls are forwarded onmodified correctly.
```
Comment (markuspf) [inline tests/Aql/ExecutionBlockImplTest.cpp]: is this supposed to be "unmodified"? Also this sounds as if an assertion is raised if the the calls are forwarded correctly.

File: tests/Aql/ExecutionBlockImplTest.cpp [hunk 6]
```diff
@@ -391,6 +401,925 @@ TEST_F(ExecutionBlockImplTest,
...
+  EXPECT_EQ(state, ExecutionState::DONE);
+  EXPECT_EQ(skipped, 1);
+  EXPECT_EQ(block, nullptr);
+  // Once with empty, once with the line by Singleton
+  EXPECT_EQ(nrCalls, 0);
+}
+
+enum CallAsserterState { INITIAL, SKIP, GET, COUNT, DONE };
+
+// Asserts if AqlCalls are forwarded onmodified correctly.
```
Comment (markuspf) [inline tests/Aql/ExecutionBlockImplTest.cpp]: is this supposed to be "unmodified"? Also this sounds as if an assertion is raised if the the calls are forwarded correctly.

File: tests/Aql/ExecutionBlockImplTest.cpp [hunk 7]
```diff
@@ -391,6 +401,1167 @@ TEST_F(ExecutionBlockImplTest,
...
+  // in the executor.
+  // Non passthrough does not make an estimate for this, so Batchsize is used.
+  ProduceCall execImpl = GetParam() ? generateProduceCall(nrCalls, fullCall, 0, 1)
+                                    : generateProduceCall(nrCalls, fullCall);
+  SkipCall skipCall = generateNeverSkipCall();
+  auto [state, skipped, block] = runTest(execImpl, skipCall, fullCall);
+
+  EXPECT_EQ(state, ExecutionState::DONE);
+  EXPECT_EQ(skipped, 0);
+  EXPECT_EQ(block, nullptr);
```
Comment (goedderz) [inline tests/Aql/ExecutionBlockImplTest.cpp]: Shouldn't this be a block with a single, empty line? I have the same question for the test cases below.

File: tests/Aql/ExecutionBlockImplTest.cpp [hunk 8]
```diff
@@ -391,6 +401,1167 @@ TEST_F(ExecutionBlockImplTest,
...
+    EXPECT_EQ(nrCalls, 2);
+  }
+
+  EXPECT_EQ(block, nullptr);
+}
+
+INSTANTIATE_TEST_CASE_P(ExecutionBlockImplExecuteTest,
+                        ExecutionBlockImplExecuteSpecificTest, ::testing::Bool());
+
+enum CallAsserterState { INITIAL, SKIP, GET, COUNT, DONE };
```
Comment (goedderz) [inline tests/Aql/ExecutionBlockImplTest.cpp]:
```suggestion
enum class CallAsserterState { INITIAL, SKIP, GET, COUNT, DONE };
```

File: tests/Aql/ExecutionBlockImplTest.cpp [hunk 9]
```diff
@@ -391,6 +401,1167 @@ TEST_F(ExecutionBlockImplTest,
...
+    // if there is neither skip nor limit
+  }
+
+  auto gotCalled(AqlCall const& got) -> void {
+    call++;
+    switch (state) {
+      case CallAsserterState::SKIP: {
+        EXPECT_EQ(got.getOffset(), expected.getOffset());
+        firstCall = !firstCall;
+        if (!firstCall) {
```
Comment (goedderz) [inline tests/Aql/ExecutionBlockImplTest.cpp]: This is confusing. `firstCall` starts as `false`. And here you first invert it then test it. What does that mean? Also, it's done in every state.

As you already have `call`, I think it'd be more readable to just check `call % 2 == 1` or so.

File: tests/Aql/ExecutionBlockImplTest.cpp [hunk 10]
```diff
@@ -391,6 +401,1167 @@ TEST_F(ExecutionBlockImplTest,
...
+};
+
+/**
+ * @brief Asserter used "above" an executor that implements
+ *        skip and produce, and transforms everything to produce.
+ *        Expects to be called twice for each sitation (with and without input).
+ *        Expect up to three situations: SKIP, GET and FULLCOUNT.
+ */
+struct GetOnlyCallAsserter : public BaseCallAsserter {
+  bool firstCall = false;
```
Comment (goedderz) [inline tests/Aql/ExecutionBlockImplTest.cpp]: Same here, I find this toggling bool confusing.

File: tests/Aql/ExecutionBlockImplTest.cpp [hunk 11]
```diff
@@ -391,6 +401,1167 @@ TEST_F(ExecutionBlockImplTest,
...
+      }
+    }
+    EXPECT_LE(call, maxCall);
+    if (call > maxCall) {
+      // Security bailout to avoid infinite loops
+      THROW_ARANGO_EXCEPTION(TRI_ERROR_INTERNAL);
+    }
+  }
+};
+
```
Comment (goedderz) [inline tests/Aql/ExecutionBlockImplTest.cpp]: Note to myself: Continue the review here.

File: arangod/Aql/ExecutionBlockImpl.cpp [hunk 12]
```diff
@@ -980,18 +1015,38 @@ std::pair<ExecutionState, SharedAqlItemBlockPtr> ExecutionBlockImpl<Executor>::r
...
 void ExecutionBlockImpl<Executor>::ensureOutputBlock(AqlCall&& call) {
-  if (_outputItemRow == nullptr || _outputItemRow->isFull()) {
+  if (_outputItemRow == nullptr || !_outputItemRow->isInitialized()) {
     _outputItemRow = allocateOutputBlock(std::move(call));
   } else {
     _outputItemRow->setCall(std::move(call));
+#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
+// We only inject a new call into the output row.
+// In the passhrough variant we need to ensure that inputBlock and outputBlock stay identical
+// TODO add an external assertion for this.
```
[the remark this answers is not in the document]
Comment (mchacki) [reply arangod/Aql/ExecutionBlockImpl.cpp]: This is in fact obsolete. solved it in the allocateOutputBlock.

File: tests/Aql/ExecutionBlockImplTest.cpp [hunk 13]
```diff
@@ -391,6 +401,1167 @@ TEST_F(ExecutionBlockImplTest,
...
+  } else {
+    EXPECT_EQ(state, ExecutionState::DONE);
+  }
+
+  ValidateResult(builder, skipped, block, outReg);
+}
+
+// Test two consecutive produce blocks.
+// The first writes 10 lines
+// The second another 100 per input (1000 in total)
```
Comment (goedderz) [inline tests/Aql/ExecutionBlockImplTest.cpp]: The comment seems misleading, doesn't it have to be
```suggestion
// The second another 10 per input (100 in total)
```
instead?

File: tests/Aql/ExecutionBlockImplTest.cpp [hunk 14]
```diff
@@ -391,6 +401,1167 @@ TEST_F(ExecutionBlockImplTest,
...
+  auto forwardCall = [&](AqlItemBlockInputRange& inputRange, OutputAqlItemRow& output)
+      -> std::tuple<ExecutorState, LambdaExe::Stats, AqlCall> {
+    while (inputRange.hasDataRow() && !output.isFull()) {
+      auto const& [state, input] = inputRange.nextDataRow();
+      EXPECT_TRUE(input.isInitialized());
+      output.copyRow(input);
+      output.advanceRow();
+    }
+    return {inputRange.upstreamState(), NoStats{}, output.getClientCall()};
+  };
```
Comment (goedderz) [inline tests/Aql/ExecutionBlockImplTest.cpp]: Both lambdas look exactly equal to me. Can you just use one?

File: arangod/Aql/ExecutionBlockImpl.cpp [hunk 15]
```diff
@@ -1061,8 +1128,12 @@ static SkipRowsRangeVariant constexpr skipRowsType() {
...
                 "Fetcher is chosen for skipping, but has not skipRows method!");
 
-  static_assert(useExecutor == (std::is_same<Executor, FilterExecutor>::value),
-                "Unexpected executor for SkipVariants::EXECUTOR");
+  static_assert(
+#ifdef ARANGODB_USE_GOOGLE_TESTS
+      useExecutor == (std::is_same_v<Executor, TestLambdaSkipExecutor>) ||
+#endif
+          useExecutor == (std::is_same_v<Executor, FilterExecutor>),
+      "Unexpected executor for SkipVariants::EXECUTOR");
```
[the remark this answers is not in the document]
Comment (mchacki) [reply arangod/Aql/ExecutionBlockImpl.cpp]: I think it works both ways, but your variant is better

[the remark this answers is not in the document]
Comment (goedderz) [reply arangod/Aql/ExecutionBlockImpl.cpp]: You're right, I made an error in thinking. Both variants are equivalent.

File: arangod/Aql/ExecutionBlockImpl.cpp [hunk 16]
```diff
@@ -1205,9 +1292,38 @@ ExecutionBlockImpl<Executor>::executeWithoutTrace(AqlCallStack stack) {
...
+          // We can either do FASTFORWARD or FULLCOUNT, difference is that
+          // fullcount counts what is produced now, FASTFORWARD simply drops
+          TRI_ASSERT(!clientCall.needsFullCount());
+          // We can drop all dataRows from upstream
+
+          while (_lastRange.hasDataRow()) {
+            auto [state, row] = _lastRange.nextDataRow();
+            TRI_ASSERT(row.isInitialized());
+          }
+          if (_lastRange.upstreamState() == ExecutorState::DONE) {
```
[the remark this answers is not in the document]
Comment (mchacki) [reply arangod/Aql/ExecutionBlockImpl.cpp]: Ok lets make it a task. Will do in Kanban

File: arangod/Aql/ExecutionBlockImpl.cpp [hunk 17]
```diff
@@ -1205,9 +1292,38 @@ ExecutionBlockImpl<Executor>::executeWithoutTrace(AqlCallStack stack) {
...
+          if (state == ExecutorState::DONE) {
+            execState = ExecState::SHADOWROWS;
+          } else {
+            // We need to request more
+            executorRequest = call;
+            execState = ExecState::UPSTREAM;
+          }
+          break;
         }
         case ExecState::UPSTREAM: {
```
[the remark this answers is not in the document]
Comment (mchacki) [reply arangod/Aql/ExecutionBlockImpl.cpp]: Yes, I fixed the assert in the ShadowRowTest Branch.
Range needs to be consumed

File: tests/Mocks/Servers.cpp [hunk 18]
```diff
@@ -450,12 +452,15 @@ std::shared_ptr<arangodb::transaction::Methods> MockAqlServer::createFakeTransac
...
 
-std::unique_ptr<arangodb::aql::Query> MockAqlServer::createFakeQuery() const {
+std::unique_ptr<arangodb::aql::Query> MockAqlServer::createFakeQuery(bool activateTracing) const {
   auto bindParams = std::make_shared<VPackBuilder>();
   bindParams->openObject();
   bindParams->close();
   auto queryOptions = std::make_shared<VPackBuilder>();
   queryOptions->openObject();
+  if (activateTracing) {
+    queryOptions->add("profile", VPackValue(5));
```
[the remark this answers is not in the document]
Comment (mchacki) [reply tests/Mocks/Servers.cpp]: Good catch

File: tests/Aql/ExecutionBlockImplTest.cpp [hunk 19]
```diff
@@ -391,6 +401,1167 @@ TEST_F(ExecutionBlockImplTest,
...
+   * @return LambdaExecutorInfos Infos to build the Executor.
+   */
+  LambdaExecutorInfos makeInfos(ProduceCall call,
+                                RegisterId inputRegisters = RegisterPlan::MaxRegisterId,
+                                RegisterId outputRegisters = RegisterPlan::MaxRegisterId) {
+    if (inputRegisters != RegisterPlan::MaxRegisterId) {
+      EXPECT_LE(inputRegisters, outputRegisters);
+    } else if (outputRegisters != RegisterPlan::MaxRegisterId) {
+      EXPECT_EQ(outputRegisters, 0);
+    }
```
[the remark this answers is not in the document]
Comment (mchacki) [reply tests/Aql/ExecutionBlockImplTest.cpp]: I agree this is confusing.
The special case in the else if is that you do not have input registers (e.g. right after singleton) but need 1 output register.
setting inputRegister == 0 means you can read from register 0 which is not available.

File: tests/Aql/ExecutionBlockImplTest.cpp [hunk 20]
```diff
@@ -391,6 +401,1167 @@ TEST_F(ExecutionBlockImplTest,
...
+    auto writeAble = make_shared_unordered_set();
+    auto registersToKeep = std::unordered_set<RegisterId>{};
+    if (inputRegisters != RegisterPlan::MaxRegisterId) {
+      for (RegisterId i = 0; i <= inputRegisters; ++i) {
+        readAble->emplace(i);
+        registersToKeep.emplace(i);
+      }
+      for (RegisterId i = inputRegisters + 1; i <= outputRegisters; ++i) {
+        writeAble->emplace(i);
+      }
```
[the remark this answers is not in the document]
Comment (mchacki) [reply tests/Aql/ExecutionBlockImplTest.cpp]: Right added assertion in the check above.

File: tests/Aql/ExecutionBlockImplTest.cpp [hunk 21]
```diff
@@ -391,6 +401,1167 @@ TEST_F(ExecutionBlockImplTest,
...
+        // This is emergency bailout, we ask way to often here
+        THROW_ARANGO_EXCEPTION(TRI_ERROR_INTERNAL);
+      }
+      nrCalls++;
+      if (input.hasDataRow()) {
+        // We expact only the empty initial row, so just consume it
+        auto const [state, row] = input.nextDataRow();
+        EXPECT_EQ(state, ExecutorState::DONE);
+        EXPECT_TRUE(row.isInitialized());
+        EXPECT_EQ(output.numRowsLeft(), numRowsLeftWithInput);
```
[the remark this answers is not in the document]
Comment (mchacki) [reply tests/Aql/ExecutionBlockImplTest.cpp]: right, fixed the comment.
The implementations are just to check the correctness of Calls

File: tests/Aql/ExecutionBlockImplTest.cpp (hunk 7 above)
Comment (mchacki) [reply tests/Aql/ExecutionBlockImplTest.cpp]: Sorry that was misleading, we do not produce rows by the above block
So we assert that we will get a nullptr back.
We assert on the Calls here, data forwarding is in integration test suite.

File: tests/Aql/ExecutionBlockImplTest.cpp (hunk 9 above)
Comment (mchacki) [reply tests/Aql/ExecutionBlockImplTest.cpp]: That is better indeed

File: tests/Aql/ExecutionBlockImplTest.cpp (hunk 14 above)
Comment (mchacki) [reply tests/Aql/ExecutionBlockImplTest.cpp]: Actually I forgot to call the state asserter, making them slightly different.
I have added refactoring in this file on the shadow row test branch
Here I fixed by calling the asserter

File: tests/Aql/ExecutionBlockImplTest.cpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (goedderz) [reply tests/Aql/ExecutionBlockImplTest.cpp]: Right, but passing a register-count instead would lead to both a consistent call and code here.

**Comment by mchacki** [comment]
Applied changes requested.

Jenkins blue:
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/8239/

File: arangod/Aql/ExecutionBlockImpl.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (goedderz) [reply arangod/Aql/ExecutionBlockImpl.cpp]:
```suggestion
```

**Comment by mchacki** [comment]
Changes addressed, got confirmation via Slack

**Pull request merged**